walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -64,6 +65,308 @@
+    }
+    if (unavailability.getTimeRange().end() == openRange.end()) {
+      uIdx++;
+    }
+    return new int[] {minUnavailability, uIdx};
+  }
+  
+  /**
+   * This method retrieves a list of the optimal free times
+   * @param openRanges is the list of free times, based on mandatory attendees
+   * @param events is all the events throughout the day
+   * @param request is the original MeetingRequest object.
+   * @return Collection<TimeRangeAndUnavailable> is a list of the optimal timeranges, and their associated unavailability of optional attendees.
+   */
+  public Collection<TimeRangeAndUnavailable> optimalTimeAndUnavailables(Collection<TimeRange> openRanges, Collection<Event> events, MeetingRequest request) {
+    List<TimeRange> openRangeList = new ArrayList(openRanges);
+    Collection<Event> optionalOnlyEvents = getOptionalOnlyEvents(events, request);
+    List<TimeRangeAndUnavailable> unavailability = new ArrayList(getAllUnavailability(optionalOnlyEvents, request));
+    Collection<TimeRangeAndUnavailable> optimalTimes = new ArrayList<>();
+
+    int uIdx = 0;
+    int minUnavailability = Integer.MAX_VALUE;
+    for (int i = 0; i < openRangeList.size(); i++) {
+      
+      //Skip all intervals that are entirely before, they're irrelevant
+      while (uIdx < unavailability.size() && !unavailability.get(uIdx).getTimeRange().overlaps(openRangeList.get(i))) {
+        uIdx++;
+      }
+      if (uIdx >= unavailability.size()) break;
+ 
+      if (isCaseOne(unavailability.get(uIdx), openRangeList.get(i), request)) {
+        int[] update = handleCaseOne(optimalTimes, unavailability.get(uIdx), openRangeList.get(i), minUnavailability, uIdx);
+        minUnavailability = update[0];
+        uIdx = update[1];
+        continue;
+      }
+
+      if (isCaseTwo(unavailability.get(uIdx), openRangeList.get(i), request)) {
+        int[] update = handleCaseTwo(optimalTimes, unavailability.get(uIdx), openRangeList.get(i), minUnavailability, uIdx);
+        minUnavailability = update[0];
+        uIdx = update[1];
+      }
+      if (uIdx >= unavailability.size()) break;
+
+      while (isCaseThree(unavailability, openRangeList.get(i), uIdx)) {
+        int[] update = handleCaseThree(optimalTimes, unavailability.get(uIdx), openRangeList.get(i), request, minUnavailability, uIdx);
+        minUnavailability = update[0];
+        uIdx = update[1];
+      }
+      if (uIdx >= unavailability.size()) break;
+
+      if (isCaseFour(unavailability.get(uIdx), openRangeList.get(i), request)) {
+        int[] update = handleCaseFour(optimalTimes, unavailability.get(uIdx), openRangeList.get(i), minUnavailability, uIdx);
+        minUnavailability = update[0];
+        uIdx = update[1];
+      }
+    }
+    return optimalTimes;
+  }
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```

Comment:
A high-level idea that I was thinking about... I believe your algorithm only returns `TimeRange`'s with a fixed # of (nonzero) available optional attendees. For example, it wouldn't return a `TimeRange` during which the # of available optional attendees changes from 2 -> 3.

Suppose, for example, that there are no mandatory attendees and you're looking for a `MeetingRequest` of duration 30 min. However, let's say the number of optional available attendees keeps increasing by one every 15 minutes. 
[0:00 - 0:15) (1 available)
[0:15 - 0:30) (2 available)
[0:30 - 0:45) (3 available)
...
[23:30 - 23:45) ("24 x 4 - 1" people available)
[23:45 - 24:00) ("24 x 4" people available)

I think your algorithm would output the empty list for this example, since each `TimeRangeAndUnavailable` is smaller than `request.duration`? However, the right solution in my mind would actually be 23:30 - 23:59. In general, the right solution would be to find a time interval that spans >= the requested duration where you attempt to "maximize the minimum # of optional attendees" at any time within the returned interval" (this is a "Minimax" formulation if you've heard the term before). Hope that sort of makes sense - this is really more of an "optional" extension to the already optional question :).

Reply:
*Edit: 23:30 - 24:00

Reply:
Thanks for this feedback! You're completely right about this and I didn't consider this case. I'm not sure if I'll be able to accomplish this in this week, as I want to direct more energy towards capstone planning and this would require me to rethink how I set up the whole problem, but I plan to go back and iterate on my solution with this feedback.

Reply:
Sounds great! Yeah, absolutely no rush at all on this - feel free to focus on the capstone project for now, and come back to it later :).